lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,215 @@
+require 'oci'
+
+module RecordStore
+  class Provider::OracleCloudDNS < Provider
+    class Error < StandardError; end
+
+    class << self
+      def client
+        config = OCI::Config.new
+        config.user = secrets["user"]
+        config.fingerprint = secrets["fingerprint"]
+        config.key_content = secrets["key_content"]
+        config.tenancy = secrets["tenancy"]
+        config.region = secrets["region"]
+        OCI::Dns::DnsClient.new(config: config)
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout) # rubocop:disable Lint/UnusedMethodArgument
+        all_records = []
+        client.get_zone_records(zone, zone_version: zone_version(zone)).each do |response|
+          all_records += response.data.items
+        end
+        all_records.map { |r| build_from_api(r) }.flatten.compact
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.list_zones(secrets["compartment_id"]).data.map(&:name)
+      end
+
+      private
+
+      # Returns the version of the zone
+      def zone_version(zone)
+        client.get_zone(zone).data.version
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        record_rdata = retrieve_rdata_based_on_record_type(record)
+
+        patch_add_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record.fqdn,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record_rdata,
+            operation: 'ADD',
+          ),
+        ]
+
+        client.patch_zone_records(
+          zone,
+          OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_add_record)
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = record.fqdn.sub(/\.$/, '') # for add, seems not require
+        record_hash = ''
+        record_rdata = retrieve_rdata_based_on_record_type(record)
+
+        client.get_zone_records(
+          zone,
+          rtype: record.type,
+          zone_version: client.get_zone(zone).data.version,
+          domain: record_fqdn,
+        ).each do |response|
+          response.data.items.each do |r|
+            record_hash = r.record_hash
```

Comment:
- Right, but the library doesn't provide matching for `rdata` in `get_zone_records`. I think they're using `zone_version` to identify properly instead of `rdata` 
[Here's the link](https://github.com/oracle/oci-ruby-sdk/blob/master/lib/oci/dns/dns_client.rb#L1188)

- You mean, is it us who generate `record_hash`? If so, It's computed by `OCI` side.

Reply:
1. What I sorta meant was that it looks like you're grabbing the `record_hash` of an arbitrary record with the `fqdn`, `type`, and, `zone` of the provided record using that loop.  This arbitrary record appears to be the last record for the `fqdn`, `type`, and `zone`.  I feel we should be searching for the `record_hash` of the record with matching `fqdn`, `type`, `zone` **and `rrdata`**.  Consider a domain with multiple TXT records...it seems like this code would always delete the last TXT record for the domain, regardless of the record provided.  But keep me honest here as I could be looking at this wrong.
2. Cool.  Yup...it's only read from OCI when zone records are fetched...gotcha.

Reply:
Updated, it will double check `rdata` is same or not.